### PR TITLE
#2764: Adding Kebabs + Delete Functionality + Modals for the Members Table and Member Profile Page - [RH]

### LINKED ISSUE #2764: Org Member Page - Three dots/ellipsis Buttons
### Issue description
A key part of member management is removing existing and invited users. To do this in our system we will have two places a user can go to remove someone, on the Member's table and in the Member profile page for a given user (invited or active). A user will click a drop down (the More actions or 3 dots icon) and this will allow someone to cancel an invitation or remove the user. 

There are certain error states where a user shouldn't be removed/uninvited, so in these cases we will still show the option to remove them but clicking it will result in an error appearing on the screen.

### Acceptance criteria

*[figma for Member Table more action buttons](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8844-31855&node-type=frame&t=33Gxdou5KBW0AmyA-0)
*[figma for Manage Member more actions](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8759-28517&node-type=frame&t=8MlasnxHsxjpy8C6-0)
- [ ] On the Member Management table3 dots/ellipsis icon shows on each member row, including rows for ones still in the invited state
- [ ] On the Member Profile table, the 3 dots/epllipsis icon matches the [Member Profile figma](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=7375-20546&node-type=frame&t=1XvlK4zEHSyP2v92-0)
 #### Complete both annotations linked above in their entirety, the following are some of the key AC in the annotation:
- [ ] Both the Member Profile page and the Member management page show the same errors when attempting to remove a user who cannot be removed
- [ ] The 3 dots drop down should result in a screen focus if it's accessed AND user can select only one of the triple dots at a time, follow the existing patterns on the Domains table
- [ ] For an **invited member**, when the 3 dots are clicked on the modal it should display "More Options > Cancel invitation" 
- [ ] For an **active member**, when the 3 dots are clicked on the modal it should display "More Options > Remove member". This should delete their UserDomainRoles only for this portfolio + remove their UserPortfolioPermissions
- [ ] If the member _manages 1 or more domains,_ follow the Figma modal "Member manages 1 or more domains" wording and display colouring
- [ ] If the member _manages no domains_, follow the Figma modal "Member manages no domains" wording and display colouring
- [ ] If the member _has an in progress request_, create an alert and follow wording and USWDS specifications per the Figma design "Member has in-progress domain request" -- in progress means anything Started, In review, Action needed.
- [ ] If the member is the only admin (see Figma for the 2 cases), and follow the Figma modal "Member is the only admin" for alert per USWDS specification + wording 
- [ ] If member is the only domain manager on a domain show the error displayed [here](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8901-71241&t=2k4IcZRT3QWTKoSC-4) if some tries to remove them.
- [ ] An admin member, who doesn't meet any of the above error states, can remove themselves from a portfolio

### Links to other issues
- [ ] Blocked by #2725 
- [ ] Blocked by #2789

### PR DESCRIPTION
**Pull request by therealslimhsiehdy**
## Ticket

Resolves #2764 

## Changes

- Adding kebab (the 3 dots) buttons on the Members Table page and on the Manage Member page (also referenced as the Member Profile page)
- Since Domain Request, Members Table and Manage Member page _almost_ all use the same kebab, table display, and modal functionality the code is made more dynamic 

## Context for reviewers
(The links are to Figmas)

- On the [Members Table page](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8844-31855), we want to add the ellipses (also referenced as kebab) buttons
- Alongside the same for the [Manage Member page which is also referenced as the Member Profile page](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8759-28517), we want to add the ellipses (also referenced as kebab) buttons
- For both we want to have options for those that are members with edit ability (Remove Member) and invited members (Cancel Invitation)
- For both we want to have modal options AND error message displays  for A) The member you're trying to delete has active requests B) The member you're trying to delete is the only admin of the org 
- We want to have a modal options AND a success message for when a user successfully deletes a member

## Setup
1. In `/admin` -> `Waffle Flags` -> turn on the flags for `organization_feature`, `organization_requests`, and `organization members`
2. Play around with domain requests from the search functionality to creating a bunch of requests to deleting requests and making sure that all runs smoothly
3. For Members specifically, there should be fixtures in place (if you run this in local and docker compose down + up it will set up the fixtures). If not, create an Organization + Suborganization and add Members to it via `User Portfolio Permissions` in Django Admin. You can invite a Member via `Portfolio Invitations` in Django Admin. 
(I'm hand waving here a bit so please ping me if it's confusing and I can help set up!)
5. Play around with deleting members, deleting invited members and testing the kebabs and modals on both the Members Table page and Member Profile/Manage Member page. 
6. You should run into these the two error scenarios and success scenarios as you try to delete members. (Feel free to ping me and I can run you through the scenarios)

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Update documentation in READMEs and/or onboarding guide

#### Ensured code standards are met (Original Developer)

- [ ] If any updated dependencies on Pipfile, also update dependencies in requirements.txt.
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] Tag @dotgov-designers in this PR's Reviewers for design review. If code is not user-facing, delete design reviewer checklist
- [ ] Verify new pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Verified code meets all checks above. Address any checks that are not satisfied
- [ ] Reviewed this code and left comments. Indicate if comments must be addressed before code is merged
- [ ] Checked that all code is adequately covered by tests
- [ ] Verify migrations are valid and do not conflict with existing migrations

#### Validated user-facing changes as a developer
**Note:** Multiple code reviewers can share the checklists above, a second reviewer should not make a duplicate checklist. All checks should be checked before approving, even those labeled N/A. 

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] (Rarely needed) Tested as both an analyst and applicant user

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior. Comment any found issues or broken flows.
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### References
- [Code review best practices](../docs/dev-practices/code_review.md)

### CODE REVIEW

**Comment by Katherine-Osos** [comment]
@therealslimhsiehdy Designers will review this function as part of a larger "design review" for member management (m5), so I removed .gov-designers as a reviewer.

File: src/registrar/assets/js/get-gov.js [hunk 1]
```diff
@@ -2105,266 +2421,68 @@ class MembersTable extends LoadTableBase {
...
+      modalHeading = `Are you sure you want to delete ${member_email}?`;
+      modalDescription = `<b>${member_email}</b> currently manages ${num_domains} domain in the organization.
+      Removing them from the organization will remove all of their domains. They will no longer be able to
+      access this organization. This action cannot be undone.`;
+    } else if (num_domains >= 1) {
+      modalHeading = `Are you sure you want to delete ${member_email}?`;
+      modalDescription = `<b>${member_email}</b> currently manages ${num_domains} domains in the organization.
+      Removing them from the organization will remove all of their domains. They will no longer be able to
+      access this organization. This action cannot be undone.`;
+    }
```
[the remark this answers is not in the document]
Comment (therealslimhsiehdy) [reply src/registrar/assets/js/get-gov.js:2449]: It's a smol thing, but it's `domain` vs `domains`. As in "currently manages 1 domain" versus "currently manages 2 domains". I did fix the second if statement to `num_domains > 1` though - good catch!

File: src/registrar/assets/js/get-gov.js [hunk 2]
```diff
@@ -86,6 +86,133 @@ function makeVisible(el) {
...
+}
+
+/**
+ * Helper function that creates a dynamic accordion navigation
+ * @param {string} action - The action type or identifier used to create a unique DOM IDs.
+ * @param {string} unique_id - An ID that when combined with action makes a unique identifier
+ * @param {string} modal_button_text - The action button's text
+ * @param {string} screen_reader_text - A screen reader helper
+ */
+function generateKebabHTML(action, unique_id, modal_button_text, screen_reader_text) {
```
[the remark this answers is not in the document]
Comment (therealslimhsiehdy) [reply src/registrar/assets/js/get-gov.js:169]: We do need `unique_id` separately in the bottom of this function so keeping separated for now!

Screenshot 2024-11-12 at 9 26 15 AM

File: src/registrar/assets/js/get-gov.js [hunk 3]
```diff
@@ -1039,29 +1167,35 @@ function initializeModals() {
...
-    this.tableWrapper = document.getElementById(`${sectionSelector}__table-wrapper`);
-    this.tableHeaders = document.querySelectorAll(`#${sectionSelector} th[data-sortable]`);
+/**
+ * Base table class which handles search, retrieval, rendering and interaction with results.
+ * Classes can extend the basic behavior of this class to customize display and interaction.
+ */
+class BaseTable {
+  constructor(itemName) {
+    this.itemName = itemName;
+    this.sectionSelector = itemName + 's';
```
[the remark this answers is not in the document]
Comment (therealslimhsiehdy) [reply src/registrar/assets/js/get-gov.js:1181]: Added a comment in the code!

File: src/registrar/assets/js/get-gov.js [hunk 4]
```diff
@@ -1903,18 +1971,186 @@ class DomainRequestsTable extends LoadTableBase {
...
+    };
+  }
+
+  addRow(dataObject, tbody, customTableOptions) {
+    const member = dataObject;
+    // member is based on either a UserPortfolioPermission or a PortfolioInvitation
+    // and also includes information from related domains; the 'id' of the org_member
+    // is the id of the UserPorfolioPermission or PortfolioInvitation, it is not a user id
+    // member.type is either invitedmember or member
+    const unique_id = member.type + member.id; // unique string for use in dom, this is
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/assets/js/get-gov.js:2064]: because you can have duplicate ids since a member can be either a permission or an invite

**Review by asaki222 — APPROVED** [review]
LGTM!

**Review by Matt-Spence — APPROVED** [review]
Great work.

**Pull request merged**